packages/babel-types/src/clone/clone.js

```diff
@@ -1,16 +1,8 @@
 // @flow
 
-/**
- * Create a shallow clone of a `node` excluding `_private` properties.
- */
-export default function clone<T: Object>(node: T): T {
-  if (!node) return node;
-  const newNode = (({}: any): T);
-
-  Object.keys(node).forEach(key => {
-    if (key[0] === "_") return;
-    newNode[key] = node[key];
-  });
+import cloneNode from "./cloneNode";
 
-  return newNode;
+export default function clone<T: Object>(node: T): T {
+  console.trace("t.clone is deprecated; use t.cloneNode instead.");
```

Comment:
I don't think deprecating these is the way to go currently. We should aim to make it as easy as possible for users to upgrade from Babel 6 to Babel 7, and while we can certainly update all of our own plugins, there are going to be third-party plugins that use `clone` and `cloneDeep`, and no user is going to upgrade to Babel 7 if it suddenly means their console is filled with deprecation warnings.

I think we probably _should_ have warnings like this in the future, but things feels like too painful for users currently. Ideally warnings should show for the people that actually have a way to address the warnings. To me that means that we should explore a way for plugin authors to opt in to additional warnings, without requiring Babel's users to have to deal with them.

For this specific case, I'm also not sure I'm comfortable with this doing a deep copy when the API says it is a shallow copy, since that could also break existing third-party code. Can we expand the API to

    export default function cloneNode<T: Object>(node: T, deep: boolean = true): T {

so it clones deeply by default but people can opt into shallow copying?

Reply:
Is there any actual use case for a shallow version of `cloneNode`? Not doing deep cloning for previously shallow `clone` seems reasonable, but for that we just have to leave the old implementation intact and be done with it.

Flooding the console with those warnings is not ideal and we should just do it once, but Im dont agree that Babel users should not see those warnings - let them see them and give them opportunity to fix those issues upstream, thats a beauty of open source, that not everything has to be on the shoulders of original creator/current maintainer.

Reply:
If, for some reason, we still need to provide a shallow clone, it'd be best to name it `shallowCloneNode` instead of just leaving `clone` as is; it's very use to misuse `clone` by accident.

Reply:
I agree with @Andarist. Most libraries show deprecation warnings even when they are not used directly, when the developer doesn't own the code which uses the deprecated API. It's a stronger motivation for the plug-in maintainers to upgrade to the new API before that it is completely removed.

I could add `grep for this message and replace console.warn with console.trace to see where it comes from` to make it easier for users to find which plugin needs to be updated (we already did something similar for a warning message I removed, I think it was the one about Flow bindings).

> For this specific case, I'm also not sure I'm comfortable with this doing a deep copy when the API says it is a shallow copy, since that could also break existing third-party code

You are right, I'll restore the old behavior of the `clone` function. Probably I should also restore the old behavior of the `cloneDeep` function, since it is an unnecessary breaking change?

> Can we expand the API to
> 
> ```js
   export default function cloneNode<T: Object>(node: T, deep: boolean = true): T {
> ```
>
> so it clones deeply by default but people can opt into shallow copying?

> If, for some reason, we still need to provide a shallow clone, it'd be best to name it shallowCloneNode

When could shallow cloning be needed?

Reply:
I have no idea, other than when the caller intentionally used `t.clone` instead of `t.cloneDeep` 🤔

Reply:
> When could shallow cloning be needed?

To me it's ideal as a performance optimization. If I want to mutate a node, but don't intend to mutate any of its children, performing a shallow clone is all that is needed. Deep being the default seems ideal, but opting in to shallow cloning does seem like a useful performance optimization to me.

> You are right, I'll restore the old behavior of the clone function. Probably I should also restore the old behavior of the cloneDeep function, since it is an unnecessary breaking change?

I don't really think it makes sense to maintain two separate copies of these APIs when it isn't absolutely required, which is why I'm in favor of just having them use your new logic.

> Flooding the console with those warnings is not ideal and we should just do it once, but Im dont agree that Babel users should not see those warnings - let them see them and give them opportunity to fix those issues upstream, thats a beauty of open source, that not everything has to be on the shoulders of original creator/current maintainer.

This seems like claiming "open source software should be painful to use", which I absolutely don't get. An average user does not care one bit that we've decided to deprecate this API, and until we actually remove it entirely, they have no reason to care. Open source means anyone can submit a fix upstream, which is fantastic, but it shouldn't mean that every user of a plugin will get an annoying message until someone submits a patch.

Ideally the behavior here would be that we have a standard plugin testbed runs a plugin against whichever versions of Babel it says it supports. Random users can submit PRs to expand the supposed Babel range, which should fail the tests if deprecated functions are used. Maintainers don't need to constantly be adjusting their Babel version but users can submit PRs, and random users don't need to worry until it actually matters to them.

Again, I'm not saying never show things to users, I'm advocating for a more stages approach. Say in a first pass we only log deprecation warnings to plugin devs, and then finally we change the function to just throw. The average user doesn't have an issue until it actually starts throwing, but devs can be aware of deprecations via their tests. We could even have a middle stage that lots the deprecation warning for all, but honestly it doesn't seem helpful.